src/java.base/share/classes/java/lang/String.java

```diff
@@ -543,48 +562,44 @@ private String(Charset charset, byte[] bytes, int offset, int length) {
                     this.coder = LATIN1;
                     return;
                 }
-                int sl = offset + length;
-                byte[] dst = new byte[length];
-                if (dp > 0) {
-                    System.arraycopy(bytes, offset, dst, 0, dp);
-                    offset += dp;
-                }
-                while (offset < sl) {
-                    int b1 = bytes[offset++];
+                // Decode with a stable copy, to be the result if the decoded length is the same
+                byte[] latin1 = Arrays.copyOfRange(bytes, offset, offset + length);
```

Comment:
This has to be moved before the `if (dp == length) { … }` check, as that also does a copy:
```java
                // Decode with a stable copy, to be the result if the decoded length is the same
                byte[] latin1 = Arrays.copyOfRange(bytes, offset, offset + length);
                int dp = StringCoding.countPositives(latin1, offset, length);
                if (dp == length) {
                    this.value = latin1;
                    this.coder = LATIN1;
                    return;
                }
```

Reply:
That may look like an improvement, to share common code, but it results in a performance hit in the normal case.
The best performing case is to copy and return immediately.